File: app/packages/core/src/resilience/simulation.ts
import type { EntityRef, SystemSchema } from '../models/schema';
import { EntityRef as EntityRefUtil } from '../models/schema';
import {
  buildResilienceRecommendations,
  resilienceRecommendationsToAdvice,
} from '../recommendations/resilienceRecommendations';
import { computeBlastRadius } from './blastRadius';
import type { ChaosSpec } from './faultSpec';
import { buildDependents, resolveFaultTargets } from './graph';
import { computeIntegrityRadius } from './integrityRadius';
import { resolveNodeResilience } from './nodeResilience';

export interface MonteCarloStats {
  iterations: number;
  overallSlaMean: number;
  overallSlaP5: number;
  overallSlaP95: number;
  entryPointSlasP95?: Record<EntityRef, number>;
}

export interface SimulationResult {
  heat: Map<EntityRef, number>;
  /** Minimum hop distance from any fault origin for animated ripple ordering. */
  heatHops: Map<EntityRef, number>;
  integrityHeat: Map<EntityRef, number>;
  impactedNodes: EntityRef[];
  integrityImpactedNodes: EntityRef[];
  entryPointSlas: Record<EntityRef, number>;
  overallSla: number;
  overallIntegrity: number;
  spofs: EntityRef[];
  impactedDomains: string[];
  integrityImpactedDomains: string[];
  advice: string[];
  propagationStoppedAt: EntityRef[];
  /** Fault injection targets resolved for this run. */
  faultNodeIds: EntityRef[];
  /** Present when WASM Monte Carlo ran, or omitted for deterministic TypeScript fallback. */
  monteCarlo?: MonteCarloStats;
  /** Which engine produced this result (`go` from WASM, `typescript` from fallback). */
  engine?: 'go' | 'typescript';
}

function resolveEntryPoints(schema: SystemSchema, explicit?: EntityRef[]): EntityRef[] {
  if (explicit && explicit.length > 0) return explicit;

  const called = new Set(schema.dependencies.map(d => d.to));
  const entryPoints = schema.nodes.map(n => n.entityRef).filter(id => !called.has(id));

  return entryPoints.length > 0 ? entryPoints : schema.nodes.map(n => n.entityRef);
}

function domainFromEntityRef(ref: EntityRef): string {
  return EntityRefUtil.getImpactedDomainGroup(ref);
}

function computeOverallIntegrity(integrityHeat: Map<EntityRef, number>): number {
  const impacted = [...integrityHeat.entries()].filter(([, intensity]) => intensity > 0);
  if (impacted.length === 0) return 100;

  const scores = impacted.map(([, intensity]) => (1 - intensity) * 100);
  const mean = scores.reduce((a, b) => a + b, 0) / scores.length;
  return Math.round(mean * 10) / 10;
}

/**
 * Detect single points of failure: dependencies with multiple callers and no circuit breaker.
 */
export function detectSpofs(schema: SystemSchema): EntityRef[] {
  const dependents = buildDependents(schema);
  const spofs: EntityRef[] = [];

  for (const [dependency, callers] of dependents) {
    if (callers.length < 2) continue;

    const node = schema.nodes.find(n => n.entityRef === dependency);
    const hasCircuitBreaker = Boolean(resolveNodeResilience(node).circuitBreaker);

    if (!hasCircuitBreaker) spofs.push(dependency);
  }

  return spofs;
}

/**
 * Deterministic resilience simulation: merge blast radii, compute SLA degradation, flag SPOFs.
 */
function mergeHeatHops(dst: Map<EntityRef, number>, src: Map<EntityRef, number>): void {
  for (const [nodeId, hop] of src) {
    const existing = dst.get(nodeId);
    dst.set(nodeId, existing == null ? hop : Math.min(existing, hop));
  }
}

/** Hop distances for ripple animation (same propagation order as blast radius). */
export function computeResilienceHeatHops(
  schema: SystemSchema,
  spec: ChaosSpec
): Map<EntityRef, number> {
  const mergedHeatHops = new Map<EntityRef, number>();

  for (const fault of spec.faults) {
    const targets = resolveFaultTargets(fault.nodeId, schema);
    for (const targetId of targets) {
      const blast = computeBlastRadius(
        schema,
        { ...fault, nodeId: targetId },
        { safeguards: spec.safeguards }
      );
      mergeHeatHops(mergedHeatHops, blast.heatHops);
    }
  }

  return mergedHeatHops;
}

/** Resolved fault injection targets for a chaos spec. */
export function resolveFaultNodeIds(schema: SystemSchema, spec: ChaosSpec): EntityRef[] {
  const faultNodeIds: EntityRef[] = [];
  for (const fault of spec.faults) {
    for (const targetId of resolveFaultTargets(fault.nodeId, schema)) {
      faultNodeIds.push(targetId);
    }
  }
  return faultNodeIds;
}

export function runResilienceSimulation(schema: SystemSchema, spec: ChaosSpec): SimulationResult {
  const mergedHeat = new Map<EntityRef, number>();
  const mergedHeatHops = new Map<EntityRef, number>();
  const mergedIntegrityHeat = new Map<EntityRef, number>();
  const propagationStoppedAt = new Set<EntityRef>();
  const faultNodeIds = resolveFaultNodeIds(schema, spec);

  for (const fault of spec.faults) {
    const targets = resolveFaultTargets(fault.nodeId, schema);
    for (const targetId of targets) {
      const blast = computeBlastRadius(
        schema,
        { ...fault, nodeId: targetId },
        {
          safeguards: spec.safeguards,
        }
      );
      for (const stopped of blast.propagationStoppedAt) propagationStoppedAt.add(stopped);
      for (const [nodeId, intensity] of blast.heat) {
        const existing = mergedHeat.get(nodeId) ?? 0;
        mergedHeat.set(nodeId, Math.min(1, Math.max(existing, intensity)));
      }
      mergeHeatHops(mergedHeatHops, blast.heatHops);

      const integrity = computeIntegrityRadius(schema, { ...fault, nodeId: targetId });
      for (const [nodeId, intensity] of integrity.integrityHeat) {
        const existing = mergedIntegrityHeat.get(nodeId) ?? 0;
        mergedIntegrityHeat.set(nodeId, Math.min(1, Math.max(existing, intensity)));
      }
    }
  }

  const entryPoints = resolveEntryPoints(schema, spec.entryPoints);
  const entryPointSlas: Record<EntityRef, number> = {};
  for (const entry of entryPoints) {
    const impact = mergedHeat.get(entry) ?? 0;
    entryPointSlas[entry] = Math.round((1 - impact) * 1000) / 10;
  }

  const slaValues = Object.values(entryPointSlas);
  const overallSla =
    slaValues.length > 0
      ? Math.round((slaValues.reduce((a, b) => a + b, 0) / slaValues.length) * 10) / 10
      : 100;

  const impactedNodes = [...mergedHeat.entries()]
    .filter(([, intensity]) => intensity > 0)
    .map(([id]) => id);

  const integrityImpactedNodes = [...mergedIntegrityHeat.entries()]
    .filter(([, intensity]) => intensity > 0)
    .map(([id]) => id);

  const impactedDomains = [...new Set(impactedNodes.map(domainFromEntityRef))];
  const integrityImpactedDomains = [...new Set(integrityImpactedNodes.map(domainFromEntityRef))];

  const spofs = detectSpofs(schema);
  const overallIntegrity = computeOverallIntegrity(mergedIntegrityHeat);

  return {
    heat: mergedHeat,
    heatHops: mergedHeatHops,
    integrityHeat: mergedIntegrityHeat,
    impactedNodes,
    integrityImpactedNodes,
    entryPointSlas,
    overallSla,
    overallIntegrity,
    spofs,
    impactedDomains,
    integrityImpactedDomains,
    advice: resilienceRecommendationsToAdvice(
      buildResilienceRecommendations({
        schema,
        spofs,
        heat: mergedHeat,
        propagationStoppedAt: [...propagationStoppedAt],
        integrityHeat: mergedIntegrityHeat,
        faultNodeIds,
      })
    ),
    propagationStoppedAt: [...propagationStoppedAt],
    faultNodeIds,
  };
}

File: resilience-engine/internal/sim/simulation.go
package sim

import (
	"math"
	"strings"

	"blueprint/resilience-engine/internal/graph"
	"blueprint/resilience-engine/internal/model"
)

func buildDependents(schema model.SystemSchema) map[string][]string {
	return graph.BuildDependents(schema)
}

func resolveEntryPoints(schema model.SystemSchema, explicit []string) []string {
	if len(explicit) > 0 {
		return explicit
	}

	called := make(map[string]struct{}, len(schema.Dependencies))
	for _, dep := range schema.Dependencies {
		called[dep.To] = struct{}{}
	}

	entryPoints := make([]string, 0)
	for _, node := range schema.Nodes {
		if _, ok := called[node.EntityRef]; !ok {
			entryPoints = append(entryPoints, node.EntityRef)
		}
	}
	if len(entryPoints) > 0 {
		return entryPoints
	}

	for _, node := range schema.Nodes {
		entryPoints = append(entryPoints, node.EntityRef)
	}
	return entryPoints
}

func domainFromEntityRef(ref string) string {
	parts := strings.Split(ref, "/")
	for _, part := range parts {
		if part != "" {
			return part
		}
	}
	return ref
}

func buildAdvice(
	schema model.SystemSchema,
	spofs []string,
	heat map[string]float64,
	integrityHeat map[string]float64,
	propagationStoppedAt []string,
	faultNodeIDs []string,
) []string {
	advice := make([]string, 0)
	nodeByID := make(map[string]model.SystemNode, len(schema.Nodes))
	for _, node := range schema.Nodes {
		nodeByID[node.EntityRef] = node
	}

	for _, spof := range spofs {
		node := nodeByID[spof]
		name := node.Name
		if name == "" {
			name = spof
		}
		advice = append(advice,
			"Add a circuit breaker on "+name+" — multiple services depend on it with no isolation.",
		)
	}

	for _, stopped := range propagationStoppedAt {
		node := nodeByID[stopped]
		name := node.Name
		if name == "" {
			name = stopped
		}
		advice = append(advice,
			"Circuit breaker on "+name+" contained the blast radius — keep this safeguard enabled.",
		)
	}

	hotNodes := make([]string, 0)
	for id, intensity := range heat {
		if intensity >= 0.7 {
			node := nodeByID[id]
			name := node.Name
			if name == "" {
				name = id
			}
			hotNodes = append(hotNodes, name)
		}
	}
	if len(hotNodes) > 0 {
		advice = append(advice, "High-impact nodes: "+strings.Join(hotNodes, ", ")+". Review timeouts and fallbacks.")
	}

	for _, faultID := range faultNodeIDs {
		brokers := graph.PubSubBrokersForPublisher(schema, faultID)
		if len(brokers) == 0 {
			continue
		}
		publisher := nodeByID[faultID]
		publisherName := publisher.Name
		if publisherName == "" {
			publisherName = faultID
		}
		stalePeers := make(map[string]struct{})
		for _, brokerID := range brokers {
			for _, peerID := range graph.PubSubPeersOnBroker(schema, brokerID) {
				if peerID == faultID {
					continue
				}
				peerHeat := integrityHeat[peerID]
				if peerHeat >= IntegrityPeerFactor*0.5 {
					stalePeers[peerID] = struct{}{}
				}
			}
		}
		if len(stalePeers) > 0 {
			peerNames := make([]string, 0, len(stalePeers))
			for id := range stalePeers {
				node := nodeByID[id]
				name := node.Name
				if name == "" {
					name = id
				}
				peerNames = append(peerNames, name)
			}
			advice = append(advice,
				publisherName+" stopped publishing — "+strings.Join(peerNames, ", ")+
					" may keep running but will miss new events.",
			)
		}
	}

	integrityOnly := make([]string, 0)
	for id, intensity := range integrityHeat {
		if intensity < 0.7 {
			continue
		}
		avail := heat[id]
		if avail >= 0.3 {
			continue
		}
		node := nodeByID[id]
		name := node.Name
		if name == "" {
			name = id
		}
		integrityOnly = append(integrityOnly, name)
	}
	if len(integrityOnly) > 0 {
		advice = append(advice,
			"Data integrity risk without availability loss: "+strings.Join(integrityOnly, ", ")+
				". Verify staleness handling and compensating actions.",
		)
	}

	return advice
}

// DetectSpofs flags shared dependencies with multiple callers and no circuit breaker.
func DetectSpofs(schema model.SystemSchema) []string {
	dependents := buildDependents(schema)
	spofs := make([]string, 0)

	for dependency, callers := range dependents {
		if len(callers) < 2 {
			continue
		}

		hasCircuitBreaker := false
		for _, node := range schema.Nodes {
			if node.EntityRef != dependency {
				continue
			}
			hasCircuitBreaker = boolVal(model.ResilienceSafeguardsForNode(node).CircuitBreaker)
			break
		}

		if !hasCircuitBreaker {
			spofs = append(spofs, dependency)
		}
	}

	return spofs
}

func mergeHeat(dst map[string]float64, src map[string]float64) {
	for id, intensity := range src {
		existing := dst[id]
		dst[id] = math.Min(1, math.Max(existing, intensity))
	}
}

func computeEntryPointSlas(
	entryPoints []string,
	heat map[string]float64,
) (map[string]float64, float64) {
	entryPointSlas := make(map[string]float64, len(entryPoints))
	values := make([]float64, 0, len(entryPoints))
	for _, entry := range entryPoints {
		impact := heat[entry]
		sla := math.Round((1-impact)*1000) / 10
		entryPointSlas[entry] = sla
		values = append(values, sla)
	}

	overall := 100.0
	if len(values) > 0 {
		sum := 0.0
		for _, v := range values {
			sum += v
		}
		overall = math.Round((sum/float64(len(values)))*10) / 10
	}

	return entryPointSlas, overall
}

// RunSimulation executes deterministic blast-radius simulation for the given spec.
func RunSimulation(schema model.SystemSchema, spec model.ChaosSpec) model.SimulationResult {
	mergedHeat := make(map[string]float64)
	mergedIntegrityHeat := make(map[string]float64)
	propagationStoppedAt := make(map[string]struct{})
	faultNodeIDs := make([]string, 0)
	options := BlastRadiusOptions{Safeguards: spec.Safeguards}

	for _, fault := range spec.Faults {
		targets := graph.ResolveFaultTargets(fault.NodeID, schema)
		for _, target := range targets {
			faultNodeIDs = append(faultNodeIDs, target)
			f := fault
			f.NodeID = target
			blast := ComputeBlastRadius(schema, f, options)
			for _, stopped := range blast.PropagationStoppedAt {
				propagationStoppedAt[stopped] = struct{}{}
			}
			mergeHeat(mergedHeat, blast.Heat)

			integrity := ComputeIntegrityRadius(schema, f)
			mergeIntegrityMaps(mergedIntegrityHeat, integrity.IntegrityHeat)
		}
	}

	entryPoints := resolveEntryPoints(schema, spec.EntryPoints)
	entryPointSlas, overallSla := computeEntryPointSlas(entryPoints, mergedHeat)

	impactedNodes := make([]string, 0)
	for id, intensity := range mergedHeat {
		if intensity > 0 {
			impactedNodes = append(impactedNodes, id)
		}
	}

	integrityImpactedNodes := make([]string, 0)
	for id, intensity := range mergedIntegrityHeat {
		if intensity > 0 {
			integrityImpactedNodes = append(integrityImpactedNodes, id)
		}
	}

	domainSet := make(map[string]struct{})
	for _, id := range impactedNodes {
		domainSet[domainFromEntityRef(id)] = struct{}{}
	}
	impactedDomains := make([]string, 0, len(domainSet))
	for domain := range domainSet {
		impactedDomains = append(impactedDomains, domain)
	}

	integrityDomainSet := make(map[string]struct{})
	for _, id := range integrityImpactedNodes {
		integrityDomainSet[domainFromEntityRef(id)] = struct{}{}
	}
	integrityImpactedDomains := make([]string, 0, len(integrityDomainSet))
	for domain := range integrityDomainSet {
		integrityImpactedDomains = append(integrityImpactedDomains, domain)
	}

	stopped := make([]string, 0, len(propagationStoppedAt))
	for id := range propagationStoppedAt {
		stopped = append(stopped, id)
	}

	spofs := DetectSpofs(schema)
	heatOut := make(map[string]float64, len(mergedHeat))
	for k, v := range mergedHeat {
		heatOut[k] = v
	}

	integrityOut := make(map[string]float64, len(mergedIntegrityHeat))
	for k, v := range mergedIntegrityHeat {
		integrityOut[k] = v
	}

	return model.SimulationResult{
		Heat:                     heatOut,
		IntegrityHeat:            integrityOut,
		ImpactedNodes:            impactedNodes,
		IntegrityImpactedNodes:   integrityImpactedNodes,
		EntryPointSlas:           entryPointSlas,
		OverallSla:               overallSla,
		OverallIntegrity:         computeOverallIntegrity(mergedIntegrityHeat),
		Spofs:                    spofs,
		ImpactedDomains:          impactedDomains,
		IntegrityImpactedDomains: integrityImpactedDomains,
		Advice: buildAdvice(
			schema,
			spofs,
			mergedHeat,
			mergedIntegrityHeat,
			stopped,
			faultNodeIDs,
		),
		PropagationStoppedAt: stopped,
		Engine:               "go",
	}
}
